Add ap capability flow to controller, agent, and CI

Galvien opened the pull request:
This pull request adds support for passing Controller Certification AP Capability Query test as defined in section 5.6.1 in the EasyMesh Test Plan v1.0 [1].

First, a shell test flow as described in [1][5.6.1] for the AP Capability Query sent by the controller to the agent is added to `test_flows.sh`.

Then, the controller is updated to support sending the query message,
Using the `WFA_CA` interface to trigger the sending of a CMDU from the controller to the agent.

Lastly, the agent backhaul manager was updated to subscribe and handle the received message, thus passing the test.

Implements #164 👌 
(A proper implementation of #161 )

Note that the current handling of the message by the agent is stub, and will be handled later during agent certification.

Commit:
tools: docker: add ap capability flow test

Add a test that triggers the sending of an AP Capability Query message
using the WFA_CA format utilized by the WFA_UCC testbed.
(see 5.6.1 in EasyMesh test plan v1.0) and checks the logs of the main
agent to see if it was handled.

While we're at it:
- Remove client/ap_capability_info_reporting from help text; they're
  called client/ap_capability_query.
- Remove the incorrectly named stub function
  ap_capability_info_reporting.
- Remove the link_metric_collection test; it doesn't exist.
- Correct the combined_infra_metrics help text.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
controller: add ap capability handling to wfa_ca

Add the creation and handling of an `ap capability query CMDU` to wfa_ca.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
agent: subscribe to and handle ap capability query

Add subscription to `AP_CAPABILITY_QUERY_MESSAGE` to the backhaul
manager thread.

The backhaul manager thread doesn't handle this message itself, which
implies that it gets forwarded to the slave threads.  Add a stub
handler for the query in the slave thread, adding a confirmation that
the message was recieved to the slave agent log.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
travis CI: add AP Capability Query test command

Add the implemented ap capability query flow test to CI.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -660,6 +660,16 @@ bool wfa_ca::create_cmdu(ieee1905_1::CmduMessageTx &cmdu_tx, ieee1905_1::eMessag
+
+        cmdu_header = cmdu_tx.create(g_mid, ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE);
+        if (!cmdu_header) {
+            //TODO: change this to the generic error message format
```
So delete the TODO...

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1581,6 +1585,11 @@ bool main_thread::handle_1905_discovery_query(ieee1905_1::CmduMessageRx &cmdu_rx
 }
+bool main_thread::handle_1905_ap_capability_query(ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    LOG(DEBUG) << "Received AP Capability Query message";
```
Add TODO comment here that explains what needs to be done - send the query to all radio agents. Each agent will then respond with its capabilities, and eventually will be sent to the controller.

morantr replied:
This is wrong, as said before, the handler should not be here.
CMDUs are passed automatically to the son_slaves.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -69,6 +69,8 @@ class main_thread : public btl::transport_socket_thread {
     bool handle_1905_autoconfiguration_response(ieee1905_1::CmduMessageRx &cmdu_rx,
                                                 const std::string &src_mac);
     bool handle_1905_discovery_query(ieee1905_1::CmduMessageRx &cmdu_rx);
+    bool handle_1905_ap_capability_query(ieee1905_1::CmduMessageRx &cmdu_rx);
```
not needed for this stage, remove

Galvien replied:
I might be missing out something but I'm pretty sure this is just the header for the `main_thread` function that we're using to log.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -476,6 +476,8 @@ bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
         return handle_channel_preference_query(sd, cmdu_rx);
     case ieee1905_1::eMessageType::CHANNEL_SELECTION_REQUEST_MESSAGE:
         return handle_channel_selection_request(sd, cmdu_rx);
+    case ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE:
```
same here

tomereli commented on agent/src/beerocks/slave/son_slave_thread.h:
```diff
@@ -283,6 +283,7 @@ class slave_thread : public beerocks::socket_thread {
     bool autoconfig_wsc_add_m1();
     bool handle_channel_preference_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
     bool handle_channel_selection_request(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
+    bool handle_1905_ap_capability_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```
and here

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -91,14 +91,12 @@ test_channel_selection() {
-test_client_capability_info_reporting() {
-    #TODO: Implement
-    return 1
+test_ap_capability_query() {
```
If you change the name, it should also be changed in the help text and in the ALL variable. But I suppose that wasn't intentional...

Galvien replied:
The ALL variable is fine but I do in fact need to fix `usage()` and I'll do that now.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -91,14 +91,12 @@ test_channel_selection() {
-    return 1
-}
-
-test_client_capability_info_reporting() {
```
Was this function removed intentionally?

Galvien replied:
Yes, the weirdly named functions were renamed to `ap_capability_query` and `client_capability_query`.

arnout reviewed (changes requested):
I can fix the issue myself, if you confirm that the rename and the removal of functions were accidental.

Galvien commented on tools/docker/tests/test_flows.sh:
```diff
@@ -102,10 +102,6 @@ test_client_capability_query() {
      #TODO: Implement
     return 1
 }
-test_link_metric_collection() {
```
This was a misnamed test, the proper test name is `combined_infra_metrics`.

Galvien commented on tools/docker/tests/test_flows.sh:
```diff
@@ -159,14 +155,12 @@ usage() {
     echo "      ap_config_bss_tear_down - AP configuration BSS Tear Down test"
     echo "      channel_selection - Channel Selection test"
-    echo "      ap_capability_info_reporting - AP Capability info reporting test"
-    echo "      client_capability_info_reporting - Capability info reporting test"
```
These two tests also don't belong

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1560,6 +1561,9 @@ bool main_thread::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
         return handle_1905_discovery_query(cmdu_rx);
         //LOG(INFO) << "I got the Topology Query message!";
     }
+    case ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE: {
```
I mention it on #161, this handler should not be here, it should be in son_slave.
It doesn't matter that it doesn't have TLVs. Channel preference query is also a message without TLVs and being handled in the son_slave which is responsible for the radio.
The backhaul manager automatically forwards messages to the son_slaves if not handled in the backhaul manager itself.
I don't see the point of handling it here and then create new CMDU to send it to the son_slave.

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4696,6 +4698,13 @@ bool slave_thread::parse_non_intel_join_response(Socket *sd)
+bool slave_thread::handle_1905_ap_capability_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received AP_CAPABILITY_QUERY_MESSAGE" << std::dec << int(mid);
```
Ok AP_CAPABILITY_QUERY_MESSAGE was received, then what?

Commit:
tools: test_flows.sh: fix redirect in client_capability_query test

We generally don't care about the output of send_bml_command, so
redirect it with $redirect like we do in the other tests.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
docker: scripts: functions.sh change success color

Change success color to green

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

tomereli reviewed (approved):
please rebase-squash then we can merge

arnout reviewed (approved):
I made a few small improvements to the commit messages.
